// @kushki/js/lib/gateway/KushkiGateway.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
exports.__esModule = true;
/**
 * Kushki Gateway File
 */
/* tslint:disable: no-unnecessary-callback-wrapper */
var AurusError_1 = require("./../../lib/generic/AurusError");
var EnvironmentEnum_1 = require("./../../lib/infrastructure/EnvironmentEnum");
var PathEnum_1 = require("./../../lib/infrastructure/PathEnum");
var inversify_1 = require("inversify");
require("reflect-metadata");
var rp = require("request-promise");
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
/**
 * Kushki Gateway Implementation
 */
var KushkiGateway = /** @class */ (function () {
    function KushkiGateway() {
        this._publicHeader = "Public-Merchant-Id";
        this._uatUrl = "";
        this._prodUrl = "";
    }
    KushkiGateway_1 = KushkiGateway;
    KushkiGateway._aurusError = function (data) {
        return data.hasOwnProperty("response_code");
    };
    KushkiGateway.prototype.request = function (body, mid, path, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(this._assignChannel(regional, path)).pipe(operators_1.switchMap(function () {
            var _a;
            return rp.post({
                body: body,
                headers: (_a = {},
                    _a[_this._publicHeader] = mid,
                    _a),
                json: true,
                uri: testEnv ? _this._uatUrl : _this._prodUrl
            });
        }), operators_1.catchError(function (err) {
            if (err.name === "StatusCodeError" &&
                KushkiGateway_1._aurusError(err.error))
                throw new AurusError_1.AurusError(err.error.response_code, err.error.response_text);
            return rxjs_1.throwError(err);
        }));
    };
    KushkiGateway.prototype.requestGet = function (path, testEnv, regional, mid) {
        var _this = this;
        return rxjs_1.of(this._assignChannel(regional, path)).pipe(operators_1.switchMap(function () {
            var _a;
            return rp.get({
                headers: (_a = {},
                    _a[_this._publicHeader] = mid,
                    _a),
                json: true,
                uri: testEnv ? _this._uatUrl : _this._prodUrl
            });
        }), operators_1.catchError(function (err) { return rxjs_1.throwError(err); }));
    };
    KushkiGateway.prototype.requestBrandsByMerchant = function (mid, testEnv, regional) {
        return this.requestGet("" + PathEnum_1.PathEnum.brands_by_merchant, testEnv, regional, mid);
    };
    KushkiGateway.prototype.requestToken = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, PathEnum_1.PathEnum.card_tokens, testEnv, regional);
        }));
    };
    KushkiGateway.prototype.requestTokenCharge = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, "" + PathEnum_1.PathEnum.card_token_charge + body.subscriptionId + PathEnum_1.PathEnum.card_token_charge, testEnv, regional);
        }));
    };
    KushkiGateway.prototype.requestSubscriptionToken = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, PathEnum_1.PathEnum.card_subscription_tokens, testEnv, regional);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.requestTransferToken = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, PathEnum_1.PathEnum.transfer_tokens, testEnv, regional);
        }));
    };
    KushkiGateway.prototype.requestMerchantSettings = function (mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.requestGet(PathEnum_1.PathEnum.merchant_settings + "settings", testEnv, regional, mid);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.requestPseBankList = function (mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.requestGet(PathEnum_1.PathEnum.bank_list, testEnv, regional, mid);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.requestDeferredConditions = function (mid, bin, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.requestGet("" + PathEnum_1.PathEnum.deferred_options + bin, testEnv, regional, mid);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.requestBinInfo = function (mid, bin, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.requestGet("" + PathEnum_1.PathEnum.bin_info + bin, testEnv, regional, mid);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.requestCashToken = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, PathEnum_1.PathEnum.cash_tokens, testEnv, regional);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.checkStatus = function (mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.requestGet("" + PathEnum_1.PathEnum.gateway_status, testEnv, regional, mid);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.requestSecureServiceValidation = function (mid, request, isTest, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(request, mid, "" + PathEnum_1.PathEnum.secure_validation, isTest, regional);
        }));
    };
    KushkiGateway.prototype.requestCardAsyncToken = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, PathEnum_1.PathEnum.card_async_tokens, testEnv, regional);
        }));
    };
    KushkiGateway.prototype.requestSubscriptionCardAsyncToken = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, PathEnum_1.PathEnum.subscription_card_async_tokens, testEnv, regional);
        }));
    };
    KushkiGateway.prototype.multiMerchantInfo = function (request, mid, isTest, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(request, mid, "" + PathEnum_1.PathEnum.multi_merchant_commission, isTest, regional);
        }));
    };
    KushkiGateway.prototype.requestPayoutsCashToken = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, PathEnum_1.PathEnum.payouts_cash_tokens, testEnv, regional);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.requestPayoutsTransferToken = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, PathEnum_1.PathEnum.payouts_transfer_tokens, testEnv, regional);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.getCommissionConfiguration = function (request, mid, isTest, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(request, mid, "" + PathEnum_1.PathEnum.commission_configuration, isTest, regional);
        }));
    };
    KushkiGateway.prototype.requestTokenTransferSubscription = function (body, mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.concatMap(function () {
            return _this.request(body, mid, PathEnum_1.PathEnum.transfer_subscription_tokens, testEnv, regional);
        }));
    };
    KushkiGateway.prototype.requestBankList = function (mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.mergeMap(function () {
            return _this.requestGet(PathEnum_1.PathEnum.bank_list_transfer_subscription, testEnv, regional, mid);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.requestPayoutsTransferBankList = function (mid, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.mergeMap(function () {
            return _this.requestGet(PathEnum_1.PathEnum.bank_list_payouts_transfer, testEnv, regional, mid);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.requestMobileProcessorToken = function (body, publicCredential, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.switchMap(function () {
            return _this.request(body, publicCredential, PathEnum_1.PathEnum.mobile_processor_token, testEnv, regional);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype.chargeMobileProcessor = function (body, publicCredential, testEnv, regional) {
        var _this = this;
        return rxjs_1.of(1).pipe(operators_1.switchMap(function () {
            return _this.request(body, publicCredential, PathEnum_1.PathEnum.mobile_processor_charge, testEnv, regional);
        }), operators_1.map(function (response) { return response; }));
    };
    KushkiGateway.prototype._assignChannel = function (regional, path) {
        if (!regional) {
            this._uatUrl = "" + EnvironmentEnum_1.EnvironmentEnum.uat + path;
            this._prodUrl = "" + EnvironmentEnum_1.EnvironmentEnum.prod + path;
        }
        else {
            this._uatUrl = "" + EnvironmentEnum_1.EnvironmentEnum.regionalUat + path;
            this._prodUrl = "" + EnvironmentEnum_1.EnvironmentEnum.regionalProd + path;
        }
        return true;
    };
    var KushkiGateway_1;
    KushkiGateway = KushkiGateway_1 = __decorate([
        inversify_1.injectable()
    ], KushkiGateway);
    return KushkiGateway;
}());
exports.KushkiGateway = KushkiGateway;
